gpu: Fix user push constants restore - and state tracking in Core validation

arno-lunarg opened the pull request:
- Push constants state restoration was not working correctly
- Reworked the state tracking, also in best practices
- Added 2 tests, `RestorePushConstants` is not working on my PC, vertex shader runs correctly but no fragment are emitted when doing an indirect draw. Things work as expected when doing a normal draw. It is not related to GPU-AV, as disabling it does not solve the problem. My indirect buffer looks ok, so I will see what CI tells me.

Will see if it closes https://github.com/KhronosGroup/Vulkan-ValidationLayers/issues/8120

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build queued with queue ID 200449.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 16826 running.

spencer-lunarg commented on tests/unit/gpu_av_positive.cpp:
```diff
@@ -1261,3 +1261,189 @@ INSTANTIATE_TEST_SUITE_P(GpuAvBufferContentValidationSettings, PositiveGpuAVPara
+
+    m_default_queue->Submit(*m_commandBuffer);
+    m_default_queue->Wait();
+}
```
I would add another test for push constant that mixes push constants for Compute and Graphics as that is a fun case we have no GPU-AV tests for

Have both push constant write data to Storage Buffer (different for compute and graphics) and make sure we didn't clobber the values by accident 

```c++
VkPipelineLayout layout_graphics; // VK_SHADER_STAGE_FRAGMENT_BIT
VkPipelineLayout layout_compute;  // VK_SHADER_STAGE_COMPUTE_BIT

// vkBeginCommandBuffer()
vkCmdBindPipeline(pipeline_graphics); // layout_graphics
vkCmdBindPipeline(pipeline_compute);  // layout_compute

vkCmdPushConstants(layout_graphics); // VK_SHADER_STAGE_FRAGMENT_BIT
vkCmdDraw();

vkCmdPushConstants(layout_compute); // VK_SHADER_STAGE_COMPUTE_BIT
vkCmdDispatch(); // valid
```

arno-lunarg replied:
yes good test case!

arno-lunarg replied:
Added it as `RestoreUserPushConstants2`

spencer-lunarg commented on layers/state_tracker/state_tracker.cpp:
```diff
@@ -2842,27 +2842,23 @@ void ValidationStateTracker::PostCallRecordCmdPushConstants(VkCommandBuffer comm
+    push_constant_data.values.resize(size);
+    auto byte_values = static_cast<const std::byte *>(pValues);
+    std::copy(byte_values, byte_values + size, push_constant_data.values.data());
+    cb_state->push_constant_data_chunks.emplace_back(push_constant_data);
```
If I go

```c++
vkCmdPushConstants(offset: 0, size: 16, value = [0, 0, 0, 0]);
vkCmdDraw(); // values = [0, 0, 0, 0]

vkCmdPushConstants(offset: 4, size: 8, value = [1 ,1]);
vkCmdDraw(); // values = [0, 1, 1, 0]
```

you shuold be overriding the values, this seem like it will make 2 copies where the first is actually outdated

arno-lunarg replied:
it will indeed not override, but there is no need to take care of this case: when restoring, I will do the exact same 2 dispatches you describe since I lookup elements in `push_constant_data_chunks` from 0 to N

spencer-lunarg replied:
1. This needs a `// comment`
2. I think this will work, the only downside is we will have to track every `vkCmdPushConstants` call made so **if** they call this 100 times in the same range, we are now tracking all that memory and have to transverse it.... I have zero knowledge how many times an app will update the `vkCmdPushConstants` for a given range, so for now happy with this approuch (but leave a comment explaining all of this for the sad soul who has to touch this code next after us 2 lol)

arno-lunarg replied:
A comment is welcome indeed.
Not to worried about not optimizing this, not doing the same call multiple times is probably more up to the application to avoid doing. And even with a 1000 elements, it is still a small number and traversing the vector should not be a problem.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 16826 failed.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build queued with queue ID 200879.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 16831 running.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 16831 failed.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build queued with queue ID 200900.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 16832 running.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 16832 failed.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build queued with queue ID 200966.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 16833 running.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 16833 passed.

spencer-lunarg commented on layers/best_practices/bp_state.h:
```diff
@@ -198,8 +198,7 @@ class CommandBuffer : public vvl::CommandBuffer {
 
-    std::vector<uint8_t> push_constant_data_set;
-    void UnbindResources() { push_constant_data_set.clear(); }
+    void UnbindResources() {}
```
why keep this?

arno-lunarg replied:
When I looked at the spot where `UnbindResources` is called, I felt that the logic was logic to call it or not was not easy to grasp, so I left this empty function to ease the process of adding such a function back if it is needed

spencer-lunarg replied:
maybe a comment... this will be strange if I came across this in the future

even a `// todo - remove` type comment

arno-lunarg replied:
done

spencer-lunarg commented on layers/state_tracker/pipeline_layout_state.h:
```diff
@@ -73,7 +73,7 @@ class PipelineLayout : public StateObject {
     const SetLayoutVector set_layouts;
     // canonical form IDs for the "compatible for set" contents
-    const PushConstantRangesId push_constant_ranges;
+    const PushConstantRangesId push_constant_ranges_layout;
```
ping on answer for this

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build queued with queue ID 201971.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 16842 running.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 16842 passed.

spencer-lunarg reviewed (changes requested):
I added the 3 positive test to ToT and ran and they pass

So I guess if we are marking this as a "clean up" then this would be ok, but we are claiming this "fixes" the push constant restore, yet we added no test to prove what this fixed (and prevent regression)

(otherwise this looks great and ready to approve and merge)

arno-lunarg commented:
> I added the 3 positive test to ToT and ran and they pass
> 
> So I guess if we are marking this as a "clean up" then this would be ok, but we are claiming this "fixes" the push constant restore, yet we added no test to prove what this fixed (and prevent regression)
> 
> (otherwise this looks great and ready to approve and merge)

The po

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build queued with queue ID 202208.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 16843 running.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 16843 passed.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build queued with queue ID 202399.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 16847 running.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 16847 passed.
